src/mlpack/core/optimizers/saga/saga_impl.hpp

```diff
@@ -0,0 +1,185 @@
+  {
+    // Calculate the objective function.
+    overallObjective = 0;
+    for (size_t f = 0; f < numFunctions; f += batchSize)
+    {
+      const size_t effectiveBatchSize = std::min(batchSize, numFunctions - f);
+      overallObjective += function.Evaluate(iterate, f, effectiveBatchSize);
+    }
+
+    if (std::isnan(overallObjective) || std::isinf(overallObjective))
+    {
+      Log::Warn << "SAGA: converged to " << overallObjective
+                << "; terminating  with failure.  Try a smaller step size?"
+                << std::endl;
+      return overallObjective;
+    }
+
+    if (std::abs(lastObjective - overallObjective) < tolerance)
+    {
+      Log::Info << "SAGA: minimized within tolerance " << tolerance
+                << "; terminating optimization." << std::endl;
+      return overallObjective;
+    }
+
+    lastObjective = overallObjective;
+
+    // initialize the gradients.
+    size_t effectiveBatchSize = std::min(batchSize, numFunctions);
+    arma::cube tableOfGradients(iterate.n_rows, iterate.n_cols, numBatches);
+    arma::mat avgGradient(iterate.n_rows, iterate.n_cols);
+
+    for (size_t f = 0, b=0; f < numFunctions;
+      /* incrementing done manually */)
+    {
+      // Find the effective batch size (the last batch may be smaller).
+      effectiveBatchSize = std::min(batchSize, numFunctions - f);
+
+      function.Gradient(iterate, f, gradient, effectiveBatchSize);
+      tableOfGradients.slice(b) = gradient;
+
+      f += effectiveBatchSize;
+      avgGradient += gradient;
+      b++;
+    }
+    avgGradient /= (double) numBatches;
+
+    // Store current parameter for the calculation of the variance reduced
+    // gradient.
+    iterate0 = iterate;
+
+    for (size_t f = 0, b=0, currentFunction = 0; f < numFunctions;
+      /* incrementing done manually */)
+    {
+      // Is this iteration the start of a sequence?
+      if ((currentFunction % numFunctions) == 0)
+      {
+        currentFunction = 0;
+
+        // Determine order of visitation.
+        if (shuffle)
+          function.Shuffle();
+      }
+
+      b = math::RandInt(0, numBatches); // Random batch selected
+      // Find the effective batch size (the last batch may be smaller).
+      if (b == numBatches-1)
+      {
+        effectiveBatchSize = numFunctions - b*batchSize;
+        currentFunction = b*batchSize;
+      }
+      else
+      {
+        effectiveBatchSize = batchSize;
+        currentFunction = b*effectiveBatchSize;
+      }
+
+
+      // Calculate the gradient of a random function.
+      function.Gradient(iterate, currentFunction, gradient,
+                        effectiveBatchSize);
+      gradient0 = tableOfGradients.slice(b);
+
+      // Use the update policy to take a step.
+      updatePolicy.Update(iterate, avgGradient, gradient, gradient0,
+                          stepSize);
+      // update average
+      avgGradient += (gradient-tableOfGradients.slice(b))/numBatches;
```

Comment:
Isn't the main difference between SAG and SAGA, that in case of SAG we take the average and for SAGA we don't?

Reply:
In both of them we take average gradients but there is a difference in their update step. In sage we divide the present and previous value of the gradient that is changed in present iteration with batchsize but in SAGA we don't divide them by batchsize..we just subtract them and then add average gradient to it. Although there's very little difference in both SAGA estimate turns out to be unbiased after convexity analysis